week-3/server.js

```diff
@@ -0,0 +1,258 @@
+    ' from products p join product_availability pa on(p.id=pa.prod_id)'+
+    "join suppliers s on (pa.supp_id=s.id) where lower(p.product_name) like '%'||$1||'%'",[searchProduct.toLowerCase()], (error, result) => {
+        res.json(result.rows);
+    })
+        else
+        pool.query('select p.product_name,pa.unit_price,s.supplier_name'+
+    ' from products p join product_availability pa on(p.id=pa.prod_id)'+
+    'join suppliers s on (pa.supp_id=s.id) ;', (error, result) => {
+        res.json(result.rows);
+    });
+
+    });
+    //new GET endpoint `/customers/:customerId`
+
+    app.get("/customers/:customerId", function(req, res) {
+        const cudtomerId=parseInt(req.params.customerId);
+        pool.query('SELECT * FROM customers where id =$1',[cudtomerId], (error, result) => {
+            res.json(result.rows[0]);
+        });
+    });
+//new POST endpoint `/customers`
+
+    app.post("/customers" ,function (req,res){
+        const newName=req.body.name;
+        const newCity=req.body.city;
+        const newAdd=req.body.address;
+        const newCountry=req.body.country;
+
+        pool.query("insert into customers (name,city ,address,country)"+
+        "values($1,$2,$3,$4) returning id",[newName ,newCity ,newAdd,newCountry],(error,result)=> {
+            if ((error)== undefined){
+                //res.send("New customer added.");
+                res.json(result.rows[0].id);
+                //res.send("New customer added.");
+            }
+            else {
+            res.status(400).send("wrong request "+error)
+}
+        });
+    });
+    
+    // new POST endpoint `/products`
+    app.post("/products" ,function (req,res){
+        const newName=req.body.name;
+
+        pool.query("insert into products (product_name)"+
+        "values($1) returning id",[newName ],(error,result)=> {
+            if ((error)== undefined){
+                //res.send("New customer added.");
+                res.json(result.rows[0].id);
+                //res.send("New customer added.");
+            }
+            else {
+            res.status(400).send("wrong request "+error)
+}
+        });
+    });
+    
+    //new POST endpoint `/availability`
+    app.post("/availability" ,function (req,res){
+        const productId=parseInt( req.body.productId);
+        const supplierId=parseInt(req.body.supplierId);
+        const price=parseInt(req.body.price);
+        if (price>0 && productIdArr.includes(productId) && supplierIdArr.includes(supplierId)){
+        pool.query("insert into product_availability "+
+        "values($1,$2,$3) ",[productId,supplierId,price ],(error,result)=> {
+            if ((error)== undefined){
+                res.send("New product_availability added.");
+               // res.json(result.rows[0]);
+                //res.send("New customer added.");
+            }
+            else {
+            res.status(400).send("wrong request "+error)
+}
+        });
+    }
```

Comment:
Your idea of creating an array up front to check if the product id exists in the product table is interesting. The model answers we have been given suggest either making a query to the database inside the function or relying on foreign key constraints. I think that relying on foreign key constraints is the most common approach - if they have been set up correctly then pool.query will return a foreign key constraint error, which you can then return in the response. But what you have done is very similar to an in-memory cache which Postgres and other databases can use. It's not a bad idea, but in a real life scenario, you would probably rely on improving the inbuilt database cache to improve performance rather than creating the in-memory array yourself.